File: ggshield/core/config/user_config.py
import json
import logging
from dataclasses import field
from pathlib import Path
from typing import Any, Dict, List, Optional, Set, Tuple
from uuid import UUID

import marshmallow_dataclass
from marshmallow import ValidationError

from ggshield.core import ui
from ggshield.core.config.utils import (
    find_global_config_path,
    find_local_config_path,
    load_yaml_dict,
    remove_common_dict_items,
    replace_dash_in_keys,
    save_yaml_dict,
    update_dict_from_other,
)
from ggshield.core.config.v1_config import convert_v1_config_dict
from ggshield.core.constants import DEFAULT_LOCAL_CONFIG_PATH
from ggshield.core.errors import ParseError, UnexpectedError, format_validation_error
from ggshield.core.types import FilteredConfig, IgnoredMatch


logger = logging.getLogger(__name__)
CURRENT_CONFIG_VERSION = 2

_IGNORE_KNOWN_SECRETS_KEY = "ignore_known_secrets"


@marshmallow_dataclass.dataclass
class SecretConfig(FilteredConfig):
    """
    Holds all user-defined secret-specific settings
    """

    show_secrets: bool = False
    ignored_detectors: Set[str] = field(default_factory=set)
    ignored_matches: List[IgnoredMatch] = field(default_factory=list)
    ignored_paths: Set[str] = field(default_factory=set)
    ignore_known_secrets: bool = False
    with_incident_details: bool = False
    # if configuration key is left unset the dashboard's remediation message is used.
    all_secrets: bool = False
    prereceive_remediation_message: str = ""
    source_uuid: Optional[UUID] = None

    def add_ignored_match(self, secret: IgnoredMatch) -> None:
        """
        Add secret to ignored_matches.
        """
        for match in self.ignored_matches:
            if match.match == secret.match:
                # take the opportunity to name the ignored match
                if not match.name:
                    match.name = secret.name
                return
        self.ignored_matches.append(secret)

    def dump_for_monitoring(self) -> str:
        return json.dumps(
            {
                "show_secrets": self.show_secrets,
                "ignored_detectors_count": len(self.ignored_detectors),
                "ignored_matches_count": len(self.ignored_matches),
                "ignored_paths_count": len(self.ignored_paths),
                "ignore_known_secrets": self.ignore_known_secrets,
                "with_incident_details": self.with_incident_details,
                "has_prereceive_remediation_message": bool(
                    self.prereceive_remediation_message
                ),
                "all_secrets": self.all_secrets,
                "source_uuid": self.source_uuid,
            }
        )


@marshmallow_dataclass.dataclass
class UserConfig(FilteredConfig):
    """
    Holds all ggshield settings defined by the user in the .gitguardian.yaml files
    (local and global).
    """

    instance: Optional[str] = None
    exit_zero: bool = False
    verbose: bool = False
    insecure: bool = False
    max_commits_for_hook: int = 50
    secret: SecretConfig = field(default_factory=SecretConfig)
    debug: bool = False

    # If we hit any deprecated syntax when loading a configuration file, we do not
    # display them directly, otherwise the messages would also be shown when running
    # `ggshield config migrate`, which would be odd.
    # Instead, we store them in this list and a result_callback() function displays
    # them when we quit. When `config migrate` runs, it clears this list, so nothing
    # gets displayed.
    deprecation_messages: List[str] = field(default_factory=list)

    def save(self, config_path: Path) -> None:
        """
        Save config to config_path
        """
        dct = self.to_config_dict()
        save_yaml_dict(dct, config_path)

    def to_config_dict(self) -> Dict[str, Any]:
        dct = self.to_dict()
        default_dct = UserConfig.from_dict({}).to_dict()

        dct = remove_common_dict_items(dct, default_dct)

        dct["version"] = CURRENT_CONFIG_VERSION

        return dct

    @classmethod
    def load(cls, config_path: Optional[Path] = None) -> Tuple["UserConfig", Path]:
        """
        Load the various user configs files to create a UserConfig object:
        - global user configuration file (in the home)
        - local user configuration file (in the repository)

        Returns a UserConfig instance, and the path where updates should be saved
        """
        deprecation_messages: List[str] = []
        if config_path:
            logger.debug("Loading custom config from %s", config_path)
            dct = _load_config_dict(config_path, deprecation_messages)
            user_config = UserConfig.from_config_dict(dct)
            user_config.deprecation_messages = deprecation_messages
            return user_config, config_path

        user_config_dict: Dict[str, Any] = {}
        global_config_path = find_global_config_path()
        if global_config_path:
            dct = _load_config_dict(global_config_path, deprecation_messages)
            update_dict_from_other(user_config_dict, dct)
            logger.debug("Loaded global config from %s", global_config_path)
        else:
            logger.debug("No global config")

        local_config_path = find_local_config_path()
        if local_config_path:
            dct = _load_config_dict(local_config_path, deprecation_messages)
            update_dict_from_other(user_config_dict, dct)
            config_path = local_config_path
            logger.debug("Loaded local config from %s", local_config_path)
        else:
            logger.debug("No local config")

        user_config = UserConfig.from_config_dict(user_config_dict)
        user_config.deprecation_messages = deprecation_messages

        if config_path is None:
            config_path = Path(DEFAULT_LOCAL_CONFIG_PATH)
        return user_config, config_path

    @staticmethod
    def from_config_dict(data: Dict[str, Any]) -> "UserConfig":
        """Create a UserConfig instance. In case of error, format it and raise
        ParseError."""
        try:
            return UserConfig.from_dict(data)
        except ValidationError as exc:
            message = format_validation_error(exc)
            raise ParseError(message) from exc


UserConfig.SCHEMA = marshmallow_dataclass.class_schema(UserConfig)(
    exclude=("deprecation_messages",)
)


def _load_config_dict(
    config_path: Path, deprecation_messages: List[str]
) -> Dict[str, Any]:
    """
    Load configuration from `config_path` as a dict.
    Appends any deprecation message regarding this file to `deprecation_messages`.
    """
    try:
        # load_yaml_dict() returns None if `config_path` does not exist. When this
        # happens, initialize the config dict to an empty configuration file,
        # but set the version to the latest version so that we don't get deprecation
        # messages.
        dct = load_yaml_dict(config_path) or {"version": CURRENT_CONFIG_VERSION}
    except ValueError as exc:
        raise ParseError(str(exc)) from exc

    dash_keys = replace_dash_in_keys(dct)

    config_version = dct.pop("version", 1)
    if config_version == 2:
        if dash_keys:
            _warn_about_dash_keys(config_path, dash_keys)
        _fix_ignore_known_secrets(dct)
        _fix_allow_self_signed(dct, config_path)
    elif config_version == 1:
        deprecation_messages.append(
            f"{config_path} uses a deprecated configuration file format."
            " Run `ggshield config migrate` to migrate it to the latest version."
        )
        dct = convert_v1_config_dict(dct, deprecation_messages)
    else:
        raise UnexpectedError(f"Don't know how to load config version {config_version}")
    return dct


def _fix_ignore_known_secrets(data: Dict[str, Any]) -> None:
    """Fix a mistake done when implementing ignore-known-secrets: since this is a secret
    specific key, it should have been stored in the "secret" mapping, not in the root
    one"""

    value = data.pop(_IGNORE_KNOWN_SECRETS_KEY, None)
    if value is None:
        # No key to fix
        return

    secret_dct = data.setdefault("secret", {})
    if _IGNORE_KNOWN_SECRETS_KEY in secret_dct:
        # Do not override if it's already there
        return
    secret_dct[_IGNORE_KNOWN_SECRETS_KEY] = value


def _fix_allow_self_signed(data: Dict[str, Any], config_path: Path) -> None:
    """Convert allow_self_signed to insecure and display a deprecation warning."""
    if insecure := data.pop("allow_self_signed", None):
        ui.display_warning(
            f"{config_path}: The 'allow_self_signed' option is deprecated. "
            "Use 'insecure: true' instead."
        )
        data["insecure"] = insecure


def _warn_about_dash_keys(config_path: Path, dash_keys: Set[str]) -> None:
    for old_key in sorted(dash_keys):
        new_key = old_key.replace("-", "_")
        ui.display_warning(
            f"{config_path}: Config key {old_key} is deprecated, use {new_key} instead."
        )

File: packages/rust-config/src/user_config.rs
//! `.gitguardian.yaml` loading. Mirrors `UserConfig.load()` (user_config.py),
//! `convert_v1_config_dict()` (v1_config.py) and the discovery helpers in
//! `ggshield/core/config/utils.py`.
//!
//! A key that would change behaviour but is NOT implemented makes the hook
//! decline and fail open with a warning. Declining scans NOTHING, silently, where
//! scanning with a rule missing scans MORE, noisily — which is why a v1 config
//! (and a versionless file IS v1) is converted rather than refused.
//!
//! Keys treated as no-ops, because they do not reach this path:
//!   secret.show_secrets           the hook always censors (`censor_match`)
//!   secret.with_incident_details  only adds a token scope to the API-key
//!                                 precheck, which this binary skips
//!   secret.prereceive_remediation_message, secret.fail_on_server_error,
//!   max_commits_for_hook, verbose, debug     other commands only

use std::collections::BTreeSet;
use std::path::{Path, PathBuf};

use serde::Deserialize;

use ggshield_common::error::Error;

pub const USER_CONFIG_FILENAMES: [&str; 3] =
    [".gitguardian", ".gitguardian.yml", ".gitguardian.yaml"];
const CURRENT_CONFIG_VERSION: i64 = 2;

#[derive(Debug, Default, Clone)]
pub struct SecretConfig {
    /// Each entry is either the sha256 "ignore sha" of a policy break or a
    /// literal secret value; `is_in_ignored_matches` accepts both.
    pub ignored_matches: Vec<String>,
    pub ignored_detectors: BTreeSet<String>,
    /// Globs, as written in the file. Compiled by `ggshield_hook::exclusion`,
    /// which also adds the default wildcards.
    pub ignored_paths: Vec<String>,
    pub ignore_known_secrets: bool,
    /// When true, ignored secrets are still reported — so they still block.
    pub all_secrets: bool,
    pub filename_only: bool,
}

#[derive(Debug, Default, Clone)]
pub struct UserConfig {
    pub instance: Option<String>,
    pub exit_zero: bool,
    /// Disables TLS certificate verification for the API client (v1
    /// `allow_self_signed`). Off by default; see `api.rs`.
    pub insecure: bool,
    pub secret: SecretConfig,
}

/// A `.gitguardian.yaml` as written, before the files are merged.
///
/// Both spellings of every key are accepted through `alias`, with one intentional
/// divergence: given the *same* key in both spellings Python keeps the underscored
/// one (`replace_dash_in_keys`), while here the last spelling in the file wins.
///
/// Scalars are `Option` so "unset" and "set to the default" stay distinct across
/// the global/local merge, and an explicit `null` reads as unset the way
/// `update_dict_from_other()` treats it.
#[derive(Debug, Default, Deserialize)]
#[serde(default)]
struct RawUserConfig {
    instance: Option<String>,
    #[serde(alias = "exit-zero")]
    exit_zero: Option<bool>,
    /// `_fix_allow_self_signed()`: the v1 spelling is the same setting.
    #[serde(alias = "allow_self_signed", alias = "allow-self-signed")]
    insecure: Option<bool>,
    /// `_fix_ignore_known_secrets()`: originally accepted at the root, where it
    /// still works unless `secret` gives its own value.
    #[serde(alias = "ignore-known-secrets")]
    ignore_known_secrets: Option<bool>,
    secret: RawSecretConfig,
}

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
struct RawSecretConfig {
    #[serde(alias = "ignored-matches")]
    ignored_matches: Vec<IgnoredMatch>,
    #[serde(alias = "ignored-detectors")]
    ignored_detectors: BTreeSet<String>,
    #[serde(alias = "ignored-paths")]
    ignored_paths: Vec<String>,
    #[serde(alias = "ignore-known-secrets")]
    ignore_known_secrets: Option<bool>,
    #[serde(alias = "all-secrets")]
    all_secrets: Option<bool>,
    #[serde(alias = "filename-only")]
    filename_only: Option<bool>,
}

/// One `secret.ignored_matches` entry. `name` is never compared, so it is not read
/// here; a missing `match` is a config error in Python too.
#[derive(Debug, Deserialize)]
struct IgnoredMatch {
    #[serde(rename = "match")]
    value: String,
}

/// A v1 (legacy) `.gitguardian.yaml`, i.e. anything without `version: 2`. The
/// deprecation messages Python prints for `all_policies` and
/// `ignore_default_excludes` are not reproduced; they change no behaviour.
#[derive(Debug, Default, Deserialize)]
#[serde(default)]
struct V1Config {
    #[serde(alias = "matches-ignore")]
    matches_ignore: Vec<V1IgnoredMatch>,
    #[serde(alias = "banlisted-detectors")]
    banlisted_detectors: BTreeSet<String>,
    #[serde(alias = "paths-ignore")]
    paths_ignore: Vec<String>,
    #[serde(alias = "api-url")]
    api_url: Option<String>,
    #[serde(alias = "allow-self-signed")]
    allow_self_signed: Option<bool>,
    #[serde(alias = "exit-zero")]
    exit_zero: Option<bool>,
    instance: Option<String>,
}

/// `_convert_matches_ignore_entry()`: v1 allowed a bare ignore sha where v2 wants
/// `{name, match}`.
#[derive(Debug, Deserialize)]
#[serde(untagged)]
enum V1IgnoredMatch {
    Sha(String),
    Entry(IgnoredMatch),
}

/// Just the `version` key, read before deciding which schema the rest follows.
#[derive(Debug, Default, Deserialize)]
struct ConfigVersion {
    #[serde(default)]
    version: Option<i64>,
}

impl From<V1Config> for RawUserConfig {
    fn from(v1: V1Config) -> Self {
        // "api_url" becomes "instance", but never overwrites an explicit one.
        let instance = v1.instance.or_else(|| {
            v1.api_url
                .as_deref()
                .map(crate::config::api_to_dashboard_url)
        });
        RawUserConfig {
            instance,
            exit_zero: v1.exit_zero,
            insecure: v1.allow_self_signed,
            ignore_known_secrets: None,
            secret: RawSecretConfig {
                ignored_matches: v1
                    .matches_ignore
                    .into_iter()
                    .map(|entry| match entry {
                        V1IgnoredMatch::Sha(sha) => IgnoredMatch { value: sha },
                        V1IgnoredMatch::Entry(entry) => entry,
                    })
                    .collect(),
                ignored_detectors: v1.banlisted_detectors,
                // `copy_if_set(secret_dct, "ignored_paths", "paths_ignore")`.
                ignored_paths: v1.paths_ignore,
                ..RawSecretConfig::default()
            },
        }
    }
}

impl RawUserConfig {
    /// `update_dict_from_other()`: ignore lists accumulate across files, scalars
    /// are overwritten by the later file, and a key the later file leaves unset
    /// keeps the earlier value.
    fn merge(&mut self, later: RawUserConfig) {
        self.instance = later.instance.or_else(|| self.instance.take());
        self.exit_zero = later.exit_zero.or(self.exit_zero);
        self.insecure = later.insecure.or(self.insecure);
        self.ignore_known_secrets = later.ignore_known_secrets.or(self.ignore_known_secrets);

        let secret = &mut self.secret;
        secret.ignored_matches.extend(later.secret.ignored_matches);
        secret
            .ignored_detectors
            .extend(later.secret.ignored_detectors);
        secret.ignored_paths.extend(later.secret.ignored_paths);
        secret.ignore_known_secrets = later
            .secret
            .ignore_known_secrets
            .or(secret.ignore_known_secrets);
        secret.all_secrets = later.secret.all_secrets.or(secret.all_secrets);
        secret.filename_only = later.secret.filename_only.or(secret.filename_only);
    }

    fn resolve(self) -> Result<UserConfig, Error> {
        Ok(UserConfig {
            instance: self
                .instance
                .map(|url| url.trim_end_matches('/').to_string()),
            exit_zero: self.exit_zero.unwrap_or_default(),
            insecure: self.insecure.unwrap_or_default(),
            secret: SecretConfig {
                ignored_matches: self
                    .secret
                    .ignored_matches
                    .into_iter()
                    .map(|entry| entry.value)
                    .collect(),
                ignored_detectors: self.secret.ignored_detectors,
                ignored_paths: self.secret.ignored_paths,
                ignore_known_secrets: self
                    .secret
                    .ignore_known_secrets
                    .or(self.ignore_known_secrets)
                    .unwrap_or_default(),
                all_secrets: self.secret.all_secrets.unwrap_or_default(),
                filename_only: self.secret.filename_only.unwrap_or_default(),
            },
        })
    }
}

pub fn home_dir() -> Option<PathBuf> {
    if let Ok(dir) = std::env::var("GG_USER_HOME_DIR") {
        return Some(PathBuf::from(dir));
    }
    dirs::home_dir()
}

fn first_existing(dir: &Path) -> Option<PathBuf> {
    USER_CONFIG_FILENAMES
        .iter()
        .map(|name| dir.join(name))
        .find(|path| path.exists())
}

pub fn global_config_path() -> Option<PathBuf> {
    first_existing(&home_dir()?)
}

/// `get_project_root_dir(Path())`: the git worktree root, or the cwd when this is
/// not a git checkout. Python shells out to `git rev-parse --show-toplevel`;
/// walking up to the nearest `.git` (a file too, as linked worktrees mark their
/// root) gives the same answer without a subprocess.
pub fn project_root() -> PathBuf {
    let cwd = std::env::current_dir().unwrap_or_else(|_| PathBuf::from("."));
    let cwd = cwd.canonicalize().unwrap_or(cwd);
    for ancestor in cwd.ancestors() {
        if ancestor.join(".git").exists() {
            return ancestor.to_path_buf();
        }
    }
    cwd
}

/// `find_local_config_path()`. Note it looks ONLY at the project root — a
/// `.gitguardian.yaml` sitting in an intermediate directory between the root and
/// the cwd is invisible to ggshield, so it must be invisible here too.
pub fn local_config_path() -> Option<PathBuf> {
    first_existing(&project_root())
}

/// An empty document deserializes as `None`, which is the default config.
fn parse_yaml<T>(raw: &str, path: &Path) -> Result<T, Error>
where
    T: Default + serde::de::DeserializeOwned,
{
    serde_yaml_ng::from_str::<Option<T>>(raw)
        .map(Option::unwrap_or_default)
        .map_err(|e| Error::fatal(format!("{}: {e}", path.display())))
}

/// `_load_config_dict()` for one file.
fn load_config_file(path: &Path) -> Result<RawUserConfig, Error> {
    let raw = std::fs::read_to_string(path)
        .map_err(|e| Error::fatal(format!("{}: {e}", path.display())))?;
    // A missing `version` means 1: Python warns and converts rather than reject,
    // so declining here would disable scanning for every pre-`version` config.
    let version = parse_yaml::<ConfigVersion>(&raw, path)?
        .version
        .unwrap_or(1);
    match version {
        CURRENT_CONFIG_VERSION => parse_yaml(&raw, path),
        1 => parse_yaml::<V1Config>(&raw, path).map(Into::into),
        // `raise UnexpectedError`, which reaches the hook's fail-open handler.
        other => Err(Error::fatal(format!(
            "Don't know how to load config version {other}"
        ))),
    }
}

/// Loads the global config then the local one, merged as Python merges them.
pub fn load() -> Result<UserConfig, Error> {
    let mut merged = RawUserConfig::default();
    for path in [global_config_path(), local_config_path()]
        .into_iter()
        .flatten()
    {
        merged.merge(load_config_file(&path)?);
    }
    merged.resolve()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parse(yaml: &str) -> Result<UserConfig, Error> {
        let dir = tempfile::tempdir().expect("tempdir");
        let path = dir.path().join(".gitguardian.yaml");
        std::fs::write(&path, yaml).expect("write");
        load_config_file(&path)?.resolve()
    }

    /// GIVEN a v2 config setting every key the hook reads
    /// WHEN it is loaded
    /// THEN each one reaches the resolved config.
    #[test]
    fn reads_every_key_that_reaches_the_hook_path() {
        let config = parse(
            "version: 2\ninstance: https://gg.example.com\nexit_zero: true\nsecret:\n  \
             ignored_detectors: [Generic Password, AWS Keys]\n  ignore_known_secrets: true\n  \
             all_secrets: true\n  filename_only: true\n  ignored_matches:\n  - match: abcdef\n    \
             name: a\n  - match: fedcba\n",
        )
        .expect("parses");
        assert_eq!(config.instance.as_deref(), Some("https://gg.example.com"));
        assert!(config.exit_zero);
        assert_eq!(config.secret.ignored_matches, vec!["abcdef", "fedcba"]);
        assert!(config.secret.ignored_detectors.contains("AWS Keys"));
        assert!(config.secret.ignore_known_secrets);
        assert!(config.secret.all_secrets);
        assert!(config.secret.filename_only);
    }

    /// GIVEN keys that do not reach the ai-hook path
    /// WHEN the config is loaded
    /// THEN they are accepted and ignored rather than rejected.
    #[test]
    fn keys_that_do_not_reach_this_path_are_accepted_and_ignored() {
        let config = parse(
            "version: 2\nsecret:\n  show_secrets: true\n  \
             with_incident_details: true\n  fail_on_server_error: false\nverbose: true\n",
        )
        .expect("parses");
        assert!(config.secret.ignored_matches.is_empty());
        assert!(!config.secret.all_secrets);
    }

    /// GIVEN `secret.ignored_paths`, in v2 and in the v1 `paths_ignore` spelling
    /// WHEN the config is loaded
    /// THEN both reach `SecretConfig`, where the hook compiles them.
    #[test]
    fn ignored_paths_are_read_from_both_schemas() {
        assert_eq!(
            parse("version: 2\nsecret:\n  ignored_paths: ['**/README.md']\n")
                .expect("parses")
                .secret
                .ignored_paths,
            vec!["**/README.md"]
        );
        assert_eq!(
            parse("version: 1\npaths-ignore:\n  - '**/*.env'\n")
                .expect("parses")
                .secret
                .ignored_paths,
            vec!["**/*.env"]
        );
    }

    /// GIVEN the dashed spelling of a key, or the root-level
    /// `ignore_known_secrets`
    /// WHEN the config is loaded
    /// THEN both are honoured.
    #[test]
    fn dashed_and_root_level_spellings_are_honoured() {
        assert!(
            parse("version: 2\nsecret:\n  ignore-known-secrets: true\n")
                .expect("parses")
                .secret
                .ignore_known_secrets
        );
        assert!(
            parse("version: 2\nignore_known_secrets: true\n")
                .expect("parses")
                .secret
                .ignore_known_secrets
        );
        assert!(
            !parse(
                "version: 2\nignore_known_secrets: true\nsecret:\n  ignore_known_secrets: false\n"
            )
            .expect("parses")
            .secret
            .ignore_known_secrets
        );
    }

    /// GIVEN a v1 config, including a versionless one
    /// WHEN it is loaded
    /// THEN it is converted, not rejected: a versionless `.gitguardian.yaml` IS v1
    /// as far as Python is concerned.
    #[test]
    fn legacy_v1_config_is_converted() {
        let config = parse(
            "matches-ignore:\n  - deadbeef\n  - name: named one\n    match: cafebabe\n\
             banlisted-detectors:\n  - AWS Keys\npaths-ignore:\n  - '**/README.md'\n\
             show_secrets: true\nexit_zero: true\n",
        )
        .expect("v1 converts");
        assert_eq!(config.secret.ignored_matches, vec!["deadbeef", "cafebabe"]);
        assert!(config.secret.ignored_detectors.contains("AWS Keys"));
        assert!(config.exit_zero);
        assert!(!config.secret.ignore_known_secrets);

        let config = parse("matches-ignore:\n  - deadbeef\n").expect("converts");
        assert_eq!(config.secret.ignored_matches, vec!["deadbeef"]);
    }

    /// GIVEN a v1 `api_url`
    /// WHEN the config is loaded
    /// THEN it becomes the dashboard `instance`, unless one is set explicitly.
    #[test]
    fn v1_api_url_becomes_the_instance() {
        let config = parse("api-url: https://onprem.example.com/exposed\n").expect("converts");
        assert_eq!(
            config.instance.as_deref(),
            Some("https://onprem.example.com")
        );
        let config = parse(
            "api-url: https://onprem.example.com/exposed\ninstance: https://other.example.com\n",
        )
        .expect("converts");
        assert_eq!(
            config.instance.as_deref(),
            Some("https://other.example.com")
        );
    }

    /// GIVEN v1 keys that only produce a deprecation message
    /// WHEN the config is loaded
    /// THEN they change nothing and do not fail.
    #[test]
    fn v1_deprecated_keys_are_ignored_without_failing() {
        let config =
            parse("all_policies: true\nignore_default_excludes: true\n").expect("converts");
        assert!(config.secret.ignored_matches.is_empty());
    }

    /// GIVEN a config version this binary does not know
    /// WHEN it is loaded
    /// THEN it is fatal rather than fail-open, matching Python's UnexpectedError.
    #[test]
    fn unknown_config_version_is_an_error_in_both_implementations() {
        let Err(Error::Fatal(msg)) = parse("version: 3\n") else {
            panic!("an unknown version must be fatal, not fail-open");
        };
        assert!(
            msg.contains("Don't know how to load config version 3"),
            "{msg}"
        );
    }

    /// GIVEN a malformed config, or one whose ignored match has no `match` key
    /// WHEN it is loaded
    /// THEN it is fatal, as it is in Python.
    #[test]
    fn a_malformed_config_is_fatal() {
        for yaml in [
            "version: 2\nsecret:\n  - [unclosed\n",
            "- just\n- a\n- list\n",
            "version: 2\nsecret:\n  ignored_matches:\n  - name: no match key\n",
            "version: 2\nsecret:\n  ignored_detectors: nope\n",
        ] {
            assert!(
                matches!(parse(yaml), Err(Error::Fatal(_))),
                "{yaml:?} must be fatal"
            );
        }
    }

    /// GIVEN `insecure` (or its v1 spelling) and `source_uuid`
    /// WHEN the config is loaded
    /// THEN `insecure` is honoured and `source_uuid` is ignored: this hook never
    /// creates incidents.
    #[test]
    fn insecure_is_honoured_and_source_uuid_is_ignored() {
        for yaml in [
            "version: 2\ninsecure: true\n",
            "version: 2\nallow_self_signed: true\n",
            "allow-self-signed: true\n",
        ] {
            assert!(parse(yaml).expect("parses").insecure, "{yaml:?}");
        }
        parse("version: 2\nsecret:\n  source_uuid: 8b7e1f1a-0000-4000-8000-000000000000\n")
            .expect("source_uuid is ignored, not declined");
    }

    /// GIVEN a global and a local config
    /// WHEN they are merged
    /// THEN ignore lists accumulate while scalars take the local value, including
    /// across a v1/v2 mix.
    #[test]
    fn merge_appends_lists_and_overwrites_scalars() {
        let of = |yaml: &str| {
            let dir = tempfile::tempdir().expect("tempdir");
            let path = dir.path().join(".gitguardian.yaml");
            std::fs::write(&path, yaml).expect("write");
            load_config_file(&path).expect("parses")
        };
        let mut merged = of(
            "version: 2\ninstance: https://global.example.com\nsecret:\n  \
                            ignored_matches:\n  - match: aaa\n  ignore_known_secrets: true\n",
        );
        merged.merge(of(
            "version: 2\ninstance: https://local.example.com\nsecret:\n  \
                         ignored_matches:\n  - match: bbb\n  all_secrets: true\n  \
                         ignore_known_secrets: false\n",
        ));
        let config = merged.resolve().expect("resolves");
        assert_eq!(
            config.instance.as_deref(),
            Some("https://local.example.com")
        );
        assert_eq!(config.secret.ignored_matches, vec!["aaa", "bbb"]);
        assert!(config.secret.all_secrets);
        // The local file wins even when it sets the default value.
        assert!(!config.secret.ignore_known_secrets);

        let mut merged = of("matches-ignore:\n  - name: fp\n    match: aaa\n");
        merged.merge(of(
            "version: 2\nsecret:\n  ignored_matches:\n  - match: bbb\n",
        ));
        assert_eq!(
            merged.resolve().expect("resolves").secret.ignored_matches,
            vec!["aaa", "bbb"]
        );
    }

    /// GIVEN an empty config file
    /// WHEN it is loaded
    /// THEN it is valid and yields defaults.
    #[test]
    fn empty_config_file_is_valid() {
        let config = parse("").expect("empty file is valid");
        assert!(config.secret.ignored_matches.is_empty());
        assert!(parse("version: 2\n").is_ok());
    }
}
